### PR TITLE
cuDNN v8.3.2

### LINKED ISSUE #34: Support CUDA 11.x on ppc64le
Blocked by https://github.com/conda-forge/nvcc-feedstock/issues/73.

### LINKED ISSUE #35: Support CUDA 11.x on aarch64
No blocker, PR welcome.

### PR DESCRIPTION
**Pull request by leofang**
Close #34. Close #35. Close #36. Close #37. Close #39.

Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by leofang** [comment]
TODO after merge: remove the [ppc64le](https://github.com/conda-forge/cudnn-feedstock/tree/ppc64le) branch.

File: recipe/conda_build_config.yaml
Comment (IvanYashchuk) [inline recipe/conda_build_config.yaml:8]: Currently, in the recipe, it is specified that building is skipped for CUDA Toolkit versions other than 11.2
```
skip: True   # [cuda_compiler_version != "11.2"]
```
So I think we don't need that many keys here?

**Review by IvanYashchuk — APPROVED** [review]

File: recipe/conda_build_config.yaml
Comment (leofang) [reply recipe/conda_build_config.yaml:8]: Yes I'll clean it up later.

Comment (leofang) [reply recipe/conda_build_config.yaml:8]: Done in 51a57ec.

**Comment by leofang** [comment]
> My hypothesis is we should copy `zlib.dll` from libzlib, rename it to `zlibwapi.dll`, and distribute it with cuDNN. But I don't have a way to verify if this would work (nor do I have bandwidth for this extra surprise).

Alternatively, we can redistribute `zlibwapi.dll` provided by NVIDIA (just like we redistribute cuDNN).

**Comment by leofang** [comment]
> Alternatively, we can redistribute zlibwapi.dll provided by NVIDIA, pretending it's part of cuDNN (just like we redistribute cuDNN).

This is done in b35e330. Let me know if it's acceptable, if not I'll revert.

File: recipe/meta.yaml
Comment (IvanYashchuk) [inline recipe/meta.yaml:23]: The [installation instructions suggest](https://docs.nvidia.com/deeplearning/cudnn/install-guide/index.html#install-zlib-windows) using this dll from the URL.
`zlib` feedstock uses a different name for the DLL - `zlib.dll`, so I think it's fine to redistribute `zLibDll/zlib123dllx64.zip` here as part of the cuDNN package.

**Comment by jakirkham** [comment]
@jaimergp do you have thoughts on Windows here?

File: recipe/meta.yaml
Comment (jaimergp) [reply recipe/meta.yaml:23]: Ideally we should build it ourselves, either as a new feedstock or a different output here.

If that fails and vendoring is the only option, the License must be included in the meta.yaml.

Comment (leofang) [reply recipe/meta.yaml:23]: @IvanYashchuk do you mind taking over and finishing the last mile for me? 🙂

Comment (IvanYashchuk) [reply recipe/meta.yaml:23]: Yes, I will finish that.

**Comment by ngam** [comment]
> @leofang anything holding this (and cudnn84xx) from being released here?

Or, sorry, @IvanYashchuk...

**Comment by ngam** [comment]
oh what a minefield lol, I am also going to say bye! (but it would be great to get this going again, thanks friends!)

**Comment by ngam** [comment]
(realistically can we drop windows and move on? maybe not... but asking as someone who hasn't used windows since the early 2000s...)

**Comment by leofang** [comment]
Now that you mentioned, it's probably worth considering. I didn't consider it because it'd be a bit annoying to add Windows when it's ready while other versions already move away.

**Comment by IvanYashchuk** [comment]
We can merge this PR once https://github.com/conda-forge/zlib-feedstock/pull/54 is merged. I'm sorry that PR was blocked on me, I just resolved the conflict and addressed review feedback there, so I hope it can be merged soon.

**Comment by leofang** [comment]
No worries Ivan, thanks for taking over 🙂

**Comment by ngam** [comment]
Thanks @IvanYashchuk, we are all here to help --- it is a community effort :)

Do you know what the procedure is like to get cudnn 84xx up here too? It seems everything is manual... but I am happy to help initiate it

**Comment by leofang** [comment]
My plan was to get this done first and then update to v8.4.x, which would be easier.

In the long run, we want the conda-forge bot to automatically pick up version updates for NVIDIA packages, and https://github.com/regro/cf-scripts/pull/1530 is the first step. But we are still working with the cuDNN team to make sure this is applicable to cuDNN (which it isn't at the moment).

**Comment by ngam** [comment]
Friends, let's skip windows...........

**Comment by ngam** [comment]
Actually you know better, I will stay out of this 😅

**Comment by leofang** [comment]
This is my thinking. If we shrink the build matrix now, and keep releasing new versions for non-Windows platforms, we're essentially leaving Windows users behind, but we originally "promised" to support (not in the literal sense, ofc). Another consideration is on the completeness: This would result in some versions of cuDNN missing on anaconda.org, but only for Windows. It'd be confusing for users to find
```
conda install -c conda-forge cudnn=8.3.2
```
works on Linux but not on Windows. Let's wait for @IvanYashchuk to make the final call.

**Comment by ngam** [comment]
> This is my thinking. If we shrink the build matrix now, and keep releasing new versions for non-Windows platforms, we're essentially leaving Windows users behind, but we originally "promised" to support (not in the literal sense, ofc). Another consideration is on the completeness: This would result in some versions of cuDNN missing on anaconda.org, but only for Windows. It'd be confusing for users to find
> 
> ```
> conda install -c conda-forge cudnn=8.3.2
> ```
> 
> works on Linux but not on Windows. Let's wait for @IvanYashchuk to make the final call.

Is it me or are the nvidia people not actually interested in actually getting the nvidia packages across the line here? 😆

**Comment by IvanYashchuk** [comment]
> > This is my thinking. If we shrink the build matrix now, and keep releasing new versions for non-Windows platforms, we're essentially leaving Windows users behind, but we originally "promised" to support (not in the literal sense, ofc). Another consideration is on the completeness: This would result in some versions of cuDNN missing on anaconda.org, but only for Windows. It'd be confusing for users to find
> > 
> > ```
> > conda install -c conda-forge cudnn=8.3.2
> > ```
> > 
> > works on Linux but not on Windows. Let's wait for @IvanYashchuk to make the final call.
> 
> Is it me or are the nvidia people not actually interested in actually getting the nvidia packages across the line here? 😆 

To be fair I joined Nvidia 3 weeks ago and all of this as for many others is all volunteering effort.

**Comment by leofang** [comment]
> To be fair I joined Nvidia 3 weeks ago and all of this as for many others is all volunteering effort.

Same here. I wasn't hired for doing conda-forge at all. It's all volunteering effort, and Ivan and I do not represent NVIDIA. Some are paid to do it, but not us.

**Comment by leofang** [comment]
If there is any concern, I am happy to communicate offline via email or other means, but I don't think continuing this discussion in this PR or any public space is appropriate, especially if it's just a tease.

**Comment by leofang** [comment]
@conda-forge-admin please rerender

**Comment by leofang** [comment]
@conda-forge-admin please rerender

**Comment by leofang** [comment]
Looks like the CDN has not been synced yet...
```
$ conda search --platform win-64 libzlib-wapi
Loading channels: done
No match found for: libzlib-wapi. Search: *libzlib-wapi*

PackagesNotFoundError: The following packages are not available from current channels:

  - libzlib-wapi
```

**Comment by IvanYashchuk** [comment]
@conda-forge-admin, please restart ci

**Comment by leofang** [comment]
I think there're some issues with CDN starting yesterday/today. I've seen other packages (ex: `libarchive`) whose recent builds are not visible to the conda/mamba solver either.

**Comment by leofang** [comment]
Thanks, John! Looks like the CDN just restored. Let's give it a shot.

@conda-forge-admin restart ci

**Comment by leofang** [comment]
@conda-forge-admin rerender

**Comment by ngam** [comment]
Thanks @leofang, this is great 🥳

**Comment by leofang** [comment]
Thanks, @IvanYashchuk and all!

**Pull request closed without merging**
